### PR TITLE
#168781676-email signup verification feature

### PR DESCRIPTION
**Pull request by ngireric123**
#### What does this PR do?

- add user verify email

#### Description of Task to be completed?

- created email verification api

#### How should this be manually tested?

- Clone the repository from [here](https://github.com/andela/caret-bn-backend)
- got to to Ft-signup-email-168781676 branch
- Add SECRET_KEY variable to .env file
- Run npm install
- run sequelize db:migrate
- run sequelize db:seed:all
- run npm run dev

- in POSTMAN got http://localhost:3000/api/v1/users/register
- go to your e-mail and click verify email button

#### Any background context you want to provide?

- N/A

#### What are the relevant pivotal tracker stories?

[#168781676](https://www.pivotaltracker.com/story/show/168781676)

#### Screenshots (if appropriate)

image

Screen Shot 2019-10-10 at 18 20 38

#### Questions:
- N/A

### CODE REVIEW

**Review by Cheza-Dzabala — APPROVED** [review]

**Review by nignanthomas — CHANGES_REQUESTED** [review]
Great job @ngireric123 
Though you need to add the env variables to **.env.example**
And I didn't see tests, do you have tests?

File: src/controllers/userController.js [hunk 1]
```diff
@@ -6,6 +9,12 @@ import strings from '../utils/stringsUtil';
 import hashPassword from '../utils/hashPassword';
 import generateToken from '../utils/generateToken';
 
+const { users } = models;
+
+dotenv.config();
+sgMail.setApiKey(process.env.SENDGRID_API_KEY);
+const { APP_URL_BACKEND } = process.env;
```
Comment (nignanthomas) [inline src/controllers/userController.js]: You should also make sure **APP_URL_BACKEND** is then heroku hostname when deployed.
And define all new env variables in **.env.example**

**Review by nignanthomas — CHANGES_REQUESTED** [review]

**Review by alainmateso — APPROVED** [review]

**Review by sabin18 — APPROVED** [review]
good job.

**Review by nignanthomas — APPROVED** [review]

File: src/controllers/userController.js [hunk 2]
```diff
@@ -6,6 +9,13 @@ import strings from '../utils/stringsUtil';
 import hashPassword from '../utils/hashPassword';
 import generateToken from '../utils/generateToken';
 
+const { users } = models;
+
+dotenv.config();
+sgMail.setApiKey(process.env.SENDGRID_API_KEY);
+const APP_URL_BACKEND = process.env.NODE_ENV === 'development' ? `http://localhost:${process.env.APPLICATION_PORT}` : 'https://caret-bn-backend.herokuapp.com';
```
Comment (nakiwuge) [inline src/controllers/userController.js]: Make the `URL` dynamic you can pick the host and protocol from the request

File: src/database/config.js [hunk 3]
```diff
@@ -6,15 +6,15 @@ module.exports = {
     password: process.env.DB_PASSWORD,
     database: process.env.DEVELOPMENT_DB,
     host: process.env.DATABASE_HOST,
-    port: process.env.DATABASE_PORT,
+    // port: process.env.DATABASE_PORT,
```
Comment (nakiwuge) [inline src/database/config.js]: remove commented code

File: src/database/config.js [hunk 4]
```diff
@@ -6,15 +6,15 @@ module.exports = {
...
     dialect: 'postgres',
   },
   test: {
     username: process.env.DB_USERNAME,
     password: process.env.DB_PASSWORD,
     database: process.env.TEST_DB,
     host: process.env.DATABASE_HOST,
-    port: process.env.DATABASE_PORT,
+    // port: process.env.DATABASE_PORT,
```
Comment (nakiwuge) [inline src/database/config.js]: remove commented code

File: src/tests/signupTest.spec.js [hunk 5]
```diff
@@ -1,8 +1,18 @@
...
 import { describe, it } from 'mocha';
+import strings from '../utils/stringsUtil';
 import app from '../index';
 import testdata from './mockdata';
+import generateToken from '../utils/generateToken';
+
+dotenv.config();
+sgMail.setApiKey(process.env.SENDGRID_API_KEY);
+
+
```
Comment (nakiwuge) [inline src/tests/signupTest.spec.js:9]: remove  this line

**Review by nakiwuge — CHANGES_REQUESTED** [review]
@ngireric123 good job however I would like you to work on the feedback below
- Remove the token from the response data that is returned to the user on signup because a user can use it to verify their email without checking their emails.
- In your pr description, you added the secret API key of sendgrid. This is very dangerous. Only share such keys on the slack channel. Please remove it

I have added some other comments on the pr.

**Review by nakiwuge — APPROVED** [review]

**Pull request merged**